### ISSUE TITLE
Select, Where, etc. should return a cached result if the input enumerable is empty

### ISSUE THREAD

**Issue by jamesqo** [dotnet/corefx#11451]
We may want to consider returning a cached result, like `Array.Empty<TSource>`, if the input enumerable is empty for certain LINQ methods. This would apply to any method that is guaranteed to have a 1:1 mapping with the input (like `Select`) or anything that needs to be smaller than the input (`Distinct`, `Where`).

**Comment by svick** [comment]
For general `IEnumerable<T>`, methods like `Select` only iterate the input *after* the method returns. By that time, it's too late to decide what to return.

For types like `List<T>` or `IList<T>`, you could check the `Count` before returning, but then that `Count` could change after `Select` is called, but before the result is iterated, so this would be invalid optimization.

For types that are guaranteed not to change the number of items (arrays, immutable collections), this could work.

**Comment by VSadov** [comment]
It seems this would be applicable only for trivial cases like empty arrays and may not compose well, unless we find a way to propagate the cached/empty result, which would add more complexity.

It does not look like this would be a big enough improvement in typical scenarios.

**Comment by jamesqo** [comment]
@VSadov, you are correct; I've learned since writing this that we can only do this for arrays because of lazy eval. However, @JonHanna and I discussed specializing `Select` / `Where` for EmptyPartition in #11840. So if we returned EmptyPartition for empty arrays, and then returned more EmptyPartitions for EmptyPartitions, then we may end up avoiding a bunch of allocations for chains of `Select` / `Where` on empty arrays and it would compose better.

No need to reopen this issue, however. I'll submit relevant fixes for this in the same PR that fixes #11840.
